subprojects/model-core/src/main/java/org/gradle/internal/instantiation/ClassGenerator.java

```diff
@@ -49,9 +49,14 @@
         T newInstance(ServiceLookup services, Instantiator nested, Object[] params) throws InvocationTargetException, IllegalAccessException, InstantiationException;
 
         /**
-         * Does this constructor use the given service?
+         * Does this constructor use the given service type?
          */
-        boolean requiresService(Class<?> serviceType);
+        boolean requiresServiceByType(Class<?> serviceType);
+
+        /**
+         * Does this constructor use a service injected via the given annotation?
+         */
```

Comment:
Yes it is. Note that this is the general infrastructure which is not necessarily bound to artifact transforms. Also note that these are the services injected into the constructor of the generated type, so services injected by type into getters would also be tracked.

We could remove this method, since for artifact transforms we are only interested if an annotation triggered an injection, and the infrastructure is not yet used anywhere else.

Reply:
To circle back: We disallow constructor injection for all the custom things we register for artifact transforms by requiring annotations to be used which can only be placed on getters.